src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1046,277 @@ begin
+  (K ∙ x) ⊔ B.orthogonal (K ∙ x) = ⊤ :=
+begin
+  refine eq_top_iff.2 (λ y _, _),
+  rw submodule.mem_sup,
+  refine ⟨(B x y * (B x x)⁻¹) • x, _, y - (B x y * (B x x)⁻¹) • x, _, _⟩,
+  { exact submodule.mem_span_singleton.2 ⟨(B x y * (B x x)⁻¹), rfl⟩ },
+  { intros z hz,
+    rcases submodule.mem_span_singleton.1 hz with ⟨μ, rfl⟩,
+    simp [is_ortho, mul_assoc, inv_mul_cancel hx, hB x] },
+  { simp }
+end
+
+/-- Given a bilinear form `B` and some `x` such that `B x x ≠ 0`, the span of the singleton of `x`
+  is complement to its orthogonal complement. -/
+lemma is_compl_singleton {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  is_compl (K ∙ x) (B.orthogonal $ K ∙ x) :=
+{ inf_le_bot := eq_bot_iff.1 $ span_inf_ortho_eq_bot hx,
+  top_le_sup := eq_top_iff.1 $ span_sup_ortho_eq_top hB hx }
+
+/-- The natural isomorphism between a singleton and the quotient by its orthogonal complement. -/
+noncomputable def quotient_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.quotient_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+/-- The natural isomorphism from the product between a singleton and its orthogonal component
+  and the whole space. -/
+noncomputable def prod_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.prod_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+lemma restrict_ortho_singleton_nondegenerate (B : bilin_form K V) (hB₁ : nondegenerate B)
+  (hB₂ : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  nondegenerate $ B.restrict $ B.orthogonal (K ∙ x) :=
+begin
+  refine λ m hm, submodule.coe_eq_zero.1 (hB₁ m.1 (λ n, _)),
+  have : n ∈ (K ∙ x) ⊔ B.orthogonal (K ∙ x) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw restrict at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+open finite_dimensional
+
+variable [finite_dimensional K V]
+
+/-- Given a nondegenerate bilinear form `B`, `B.to_dual` is the isomorphism
+between a vector space and its dual with the underlying linear map `B.to_lin`. -/
+noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
+  V ≃ₗ[K] module.dual K V :=
+B.to_lin.linear_equiv_of_ker_eq_bot
+  (nondegenerate_iff_ker_eq_bot.mp hB) subspace.dual_findim_eq.symm
+
+lemma to_dual_def {B : bilin_form K V} (hB : B.nondegenerate) {m n : V} :
+  B.to_dual hB m n = B m n := rfl
+
+-- We start proving that symmetric nondegenerate bilinear forms are diagonalisable, or equivalently
+-- there exists a orthogonal basis with respect to any symmetric nondegenerate bilinear form.
+
+lemma findim_ortho_span_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B)
+  {x : V} (hx : ¬ B.is_ortho x x) :
+  findim K V = findim K (B.orthogonal $ K ∙ x) + 1 :=
```

Comment:
This lemma and `quotient_equiv_of_ortho_singleton` are a bit too specific for my taste.  Here are some more general lemmas that it would be good to have:
(1) a lemma in `linear_algebra.finite_dimensional` saying that if `U, W` are submodules and `is_compl U W` then `findim K V = findim K U + findim K W`.  (Should follow from `submodule.dim_sup_add_dim_inf_eq`.)
(2) a lemma in this file that if `¬ B.is_ortho x x` then `x ≠ 0`

Can you try deleting the two lemmas I mentioned and adding the two I mentioned?  If the use case at line 1261,
```lean
rw findim_ortho_span_singleton hB₂ hx at hd,
```
feels too clunky after the change, you could also add a lemma saying that if `x ≠ 0` and `is_compl (K ∙ x) W`
then `findim K V = findim K W + 1` -- still pretty specific, but less so.

Reply:
Done!